File: intake/src/incident-fields.js
import catalog from './catalog.json';
import priorityPolicy from './priority-matrix.json';

export const levels = priorityPolicy.levels;
const impactLabels = {
  highest: 'Major / Widespread', high: 'Significant / Large', medium: 'Moderate / Limited',
  low: 'Minor / Localized', lowest: 'No Direct Impact / Information',
};
const titleCase = value => value ? value[0].toUpperCase() + value.slice(1) : '';
export const chipFields = {
  workType: { label: 'Work Type', options: ['Incident', 'Service Request'] },
  urgency: { label: 'Urgency', options: levels, format: titleCase },
  impact: { label: 'Impact', options: levels, format: value => impactLabels[value] || value },
  priority: { label: 'Priority', kind: 'priority', format: titleCase },
  department: { label: 'Service Teams', options: catalog['Service Team(s)'] },
};

export function calculatePriority(urgency, impact) {
  const row = levels.indexOf(urgency);
  const column = levels.indexOf(impact);
  return row < 0 || column < 0 ? '' : priorityPolicy.matrix[row][column];
}

export function resolvedFields(details, inferred) {
  return { ...inferred, ...details };
}

File: intake/incident_fields.py
"""The challenge fields, their allowed values, and deterministic priority policy."""
import json
from pathlib import Path

SOURCE = Path(__file__).parent / 'src'
CATALOG = json.loads((SOURCE / 'catalog.json').read_text())
ASSIGNEES = json.loads((SOURCE / 'assignees.json').read_text())
PRIORITY = json.loads((SOURCE / 'priority-matrix.json').read_text())
LEVELS, MATRIX = PRIORITY['levels'], PRIORITY['matrix']
RESOLUTIONS = ['done', 'cancelled', 'clarification', 'cannot reproduce']
WORK_TYPES = ['Incident', 'Service Request']
CHIP_OPTIONS = {
    'workType': ('Work Type', WORK_TYPES),
    'urgency': ('Urgency', LEVELS),
    'impact': ('Impact', LEVELS),
    'priority': ('Priority', LEVELS),
    'department': ('Service Teams', CATALOG['Service Team(s)']),
}


def calculate_priority(urgency, impact):
    if urgency not in LEVELS or impact not in LEVELS:
        return None
    return MATRIX[LEVELS.index(urgency)][LEVELS.index(impact)]
